REFACTOR: Enhance Contact Form Validation

Yura33-dev opened the pull request:
## Description
This PR enhances live form validation to provide users with clearer feedback on invalid inputs. It also implements the useReducer hook to manage form state more effectively.

## Changes Made
- Expanded validation to handle additional form errors as specified in the [Project Requirements](https://docs.google.com/document/d/10mjhJZ6npoHybsiLqjVz8z4SenT9OTyt6Oc8dCOFKHY/edit#heading=h.kctencwjew16).
- Utilized React's `useReducer` hook for managing form state, resulting in improved codebase readability and maintainability.
- Added error messages to clearly indicate form submission failures.
- Ensured that error messages are cleared upon new submission attempts to provide a smoother user experience.

Yura33-dev committed:
refactor: implements useReducer hook in form validation

naumch1k committed:
refactor: restructures form reducer for better state management and code clarity

naumch1k commented:
Hey @Yura33-dev 👋

Thank you so much for setting up the form reducer! I really appreciate your work on this ❤️

I've reviewed your reducer logic and made some revisions to simplify state management and improve the overall structure for better maintainability and readability.

Here are the key takeaways:

First off, **I consolidated the field-related state into a single fields object.** Now, each field encapsulates its `value`, `error`, and `wasChanged` status. This makes the state more organized and easier to manage.

```tsx
type FormFieldState<T> = {
  [K in keyof T]: {
    value: T[K]
    wasChanged: boolean
    error?: string
  }
}
```

**Simplifying Form Validity Computation.** Since all fields are required, our form will be valid when the user has changed each of them and none of the fields have errors. Although your solution for computing `formIsValid` works well, it contains repetitive logic and can be a bit challenging to follow. To disable the submit button on the initial render and when the form resets, we can now check if each field has been changed instead of comparing each value to an empty string.

Original Logic:
```tsx
Object.values(state.errors).every(error => !error) &&
  state.values.name.trim() !== '' &&
  state.values.email.trim() !== '' &&
  state.values.message.trim() !== ''
```

Updated Logic:
```tsx
const isValid = useMemo(() => {
  return Object.values(state.fields).every(
    ({ wasChanged, error }) => wasChanged && !error
  );
}, [state.fields]);
```

Note that I've removed `isValid` from our form state and leveraged React's memoization to determine form validity whenever the fields state changes. This simplifies our logic and avoids redundant state updates.

Next, **I revised the reducer and consolidated actions** to streamline the handling of essential state transitions. Instead of focusing on granular state updates, we can now dispatch key form changes: field updates, form submission start, and form submission success or failure.

Here are the updated action types:
```tsx
| { type: 'UPDATE_FIELD'; payload: { field: K; value: ContactFormFields[K]; error?: string } }
| { type: 'START_FORM_SUBMISSION' }
| { type: 'SET_FORM_SUBMISSION_SUCCESS' }
| { type: 'SET_FORM_SUBMISSION_FAILURE'; payload: { error: string } }
| { type: 'COMPLETE_FORM_SUBMISSION' }
```

This approach allows us to avoid dispatching unnecessary payloads when we don't need them.

For example, we only reset the form when it has been successfully submitted. To do that, we reset the fields to their initial values, set `isSubmitting` to `false`, and set `formSuccessfullySent` to `true` to indicate that the form was indeed sent. Since form validity is now computed on-the-fly, we don't have to manage it in the state 😉

Here's the reducer case:
```tsx
case 'SET_FORM_SUBMISSION_SUCCESS':
  return {
    ...state,
    fields: initialContactFormFieldsState,
    isSubmitting: false,
    formSuccessfullySent: true,
  };
  ```
  
I have also moved the timeout that resets the success button message into a `useEffect` hook. This ensures it is properly cleaned up if the component unmounts, preventing memory leaks and unintended behavior.

Let me know what you think about these changes! I'm happy to discuss any part of it 🙌  I'll be adding a few more comments soon with some additional improvements we can consider.

naumch1k commented on src/helpers/hooks/useFormWithValidation.ts:
```diff
@@ -1,77 +1,192 @@
+  submissionError?: string
+}
+
+type ContactFormAction<K extends keyof ContactFormFields = keyof ContactFormFields> =
```
@Yura33-dev I have another suggestion to further improve our reducer: we can leverage [TypeScript enums](https://www.typescriptlang.org/docs/handbook/enums.html) to define our action types. This can enhance type safety, reducing the risk of typos or incorrect strings.

We can define an enum for our action types:
```tsx
enum ContactFormActionTypes {
  UpdateField,
  StartFormSubmission,
  SetFormSubmissionSuccess,
  SetFormSubmissionFailure,
  CompleteFormSubmission,
}
```

To avoid repeating `ContactFormActionTypes` every time, we can destructure the enum:
```tsx
const {
  UpdateField,
  StartFormSubmission,
  SetFormSubmissionSuccess,
  SetFormSubmissionFailure,
  CompleteFormSubmission,
} = ContactFormActionTypes
```

### Updating Action Types
Here's how we can update our `ContactFormAction` type to use the enum. We use `typeof UpdateField` in our action types to tell TypeScript that the type property must exactly match the `UpdateField` value from our enum.
```tsx
type ContactFormAction<K extends keyof ContactFormFields = keyof ContactFormFields> =
  | { type: typeof UpdateField; payload: { field: K; value: ContactFormFields[K]; error?: string } }
  | { type: typeof StartFormSubmission }
  ...
```

### Using Enums in the Reducer
In the reducer, we can now switch on the enum values:
```tsx
const reducer = (state: ContactFormState, action: ContactFormAction) => {
  switch (action.type) {
    case UpdateField:
      // Handle update field action
    case StartFormSubmission:
      // Handle start form submission action
      ...
```

### Dispatching actions
That requre payload:
```tsx
    dispatch({
      type: UpdateField,
      payload: {
        field,
        value,
        error: getFieldError(field, value),
      },
    })
```
Or as simple as:
```tsx
    dispatch({ type: StartFormSubmission })
```

naumch1k commented on src/helpers/hooks/useFormWithValidation.ts:
```diff
@@ -1,77 +1,192 @@
-    setValues({ ...values, [name]: value })
-    setErrors({ ...errors, [name]: target.validationMessage })
-    setIsValid(target.closest('form')!.checkValidity())
+  const [state, dispatch] = useReducer(reducer, initialContactFormState)
```
@Yura33-dev Let's also destructure the `state` directly when we call `useReducer`. This way, we avoid repeating state every time we access its properties.

Current Code:
```tsx
const [state, dispatch] = useReducer(reducer, initialContactFormState);

// Accessing state properties:
state.fields
state.isSubmitting
state.formSuccessfullySent
```

Proposed Change:
```tsx
const [
  { fields, isSubmitting, formSuccessfullySent },
  dispatch,
] = useReducer(reducer, initialContactFormState)
```

Now, we can use `fields`, `isSubmitting`, and `formSuccessfullySent` directly without prefixing them with state.

naumch1k commented on src/helpers/hooks/useFormWithValidation.ts:
```diff
@@ -1,77 +1,192 @@
+        break
+      case 'email':
+        if (!trimmedValue) return email.VALUE_MISSING
+        if (!/^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.(com|net)$/.test(trimmedValue)) return email.PATTERN_MISMATCH
```
@Yura33-dev I suggest we improve our email validation by:

- **Removing the `com|net` Limitation.** The current regex only allows emails ending with `.com` or `.net`, which excludes many valid domains like `.org`, `.edu`, country codes (e.g., `.uk`, `.ca`), and newer TLDs like `.io`, `.dev`, etc. Please modify the regex to accept a wider range of valid email domains.

- **Moving the Regex to a Separate File.** Please store the email regex as `validEmailRegexp` in `regexps.ts` file under `/helpers/constants` for better maintainability.

naumch1k commented on src/helpers/hooks/useFormWithValidation.ts:
```diff
@@ -1,77 +1,192 @@
-    setIsValid(target.closest('form')!.checkValidity())
+  const [state, dispatch] = useReducer(reducer, initialContactFormState)
+
+  const getFieldError = <K extends keyof ContactFormFields>(
```
@Yura33-dev As per the [Project Requirements](https://docs.google.com/document/d/10mjhJZ6npoHybsiLqjVz8z4SenT9OTyt6Oc8dCOFKHY/edit), we also need to ensure that the `name` and `message` fields do not contain special characters like `<`, `>`, `&`, which can be interpreted as part of HTML code or JavaScript. This is important for preventing potential security vulnerabilities like injection attacks.

Please setup a another regex for that. We'll need to add `PATTERN_MISMATCH` messages for both the `name` and `message` fields in `CustomValidationMessages` to provide users with clear error message. For a name can use something like:
> Only latin character, spaces, apostrophes ('), and hyphens (-) allowed

Please check with QA if you have any clarification on that 🙂

naumch1k replied:
Hi @Yura33-dev 🖐️ Thanks for adding pattern validation to the name and message fields! Would you consider these pattern mismatch error messages?

Name Field:
> Name can only include letters, spaces, and these: ' ’ -

Message Field:
> Yikes! Only letters, numbers, spaces, and ' ’ " , . ! ? - are allowed

Additionally, could you please check the message regex to make sure comma is listed only once?

naumch1k replied:
@Yura33-dev Did you see the message above?

Yura33-dev replied:
@naumch1k Yeep, I saw those messages a little later 🙈. But now I have added new commit with these enhancements 🙂

naumch1k replied:
@Yura33-dev I am sorry if I caused any confusion. I didn't mean to restrict multiple consecutive commas in `validMessageRegexp`. What I wanted was just to clean up a duplicate. 
<img width="680" alt="Screenshot 2024-09-30 at 11 49 34 AM" src="https://github.com/user-attachments/assets/ad8168a7-348d-4b44-a499-d2e10c730245">


Could you please explain the change you made to `validNameRegexp`? Particularly, the need in `z` after `-`?

Yura33-dev replied:
@naumch1k Ops! Okay, I have deleted one comma from `validMessageRegexp`. 
I suggest it looks like wrong (I mean `validNameRegexp` with `z`). I just deleted it from regexp. 🙂
I have pushed a new commit

naumch1k reviewed (dismissed):
@Yura33-dev Below are a few more suggestions to further enhance our form reducer and hook.

To keep the version history clean, could you please introduce these changes in three separate commits:
- Commit 1: Implement enums for action types.
- Commit 2: Destructure state in useFormWithValidation.
- Commit 3: Update fields validation (both email regex update and special characters restriction).

This way, each commit will have a clear purpose, making it easier to review and maintain 🤩

Definitely let me know if you have any questions or doubts!

Yura33-dev commented:
> Hey @Yura33-dev 👋
> 
> Thank you so much for setting up the form reducer! I really appreciate your work on this ❤️
> 
> I've reviewed your reducer logic and made some revisions to simplify state management and improve the overall structure for better maintainability and readability.
> 
> Here are the key takeaways:
> 
> First off, **I consolidated the field-related state into a single fields object.** Now, each field encapsulates its `value`, `error`, and `wasChanged` status. This makes the state more organized and easier to manage.
> 
> ```tsx
> type FormFieldState<T> = {
>   [K in keyof T]: {
>     value: T[K]
>     wasChanged: boolean
>     error?: string
>   }
> }
> ```
> 
> **Simplifying Form Validity Computation.** Since all fields are required, our form will be valid when the user has changed each of them and none of the fields have errors. Although your solution for computing `formIsValid` works well, it contains repetitive logic and can be a bit challenging to follow. To disable the submit button on the initial render and when the form resets, we can now check if each field has been changed instead of comparing each value to an empty string.
> 
> Original Logic:
> 
> ```tsx
> Object.values(state.errors).every(error => !error) &&
>   state.values.name.trim() !== '' &&
>   state.values.email.trim() !== '' &&
>   state.values.message.trim() !== ''
> ```
> 
> Updated Logic:
> 
> ```tsx
> const isValid = useMemo(() => {
>   return Object.values(state.fields).every(
>     ({ wasChanged, error }) => wasChanged && !error
>   );
> }, [state.fields]);
> ```
> 
> Note that I've removed `isValid` from our form state and leveraged React's memoization to determine form validity whenever the fields state changes. This simplifies our logic and avoids redundant state updates.
> 
> Next, **I revised the reducer and consolidated actions** to streamline the handling of essential state transitions. Instead of focusing on granular state updates, we can now dispatch key form changes: field updates, form submission start, and form submission success or failure.
> 
> Here are the updated action types:
> 
> ```tsx
> | { type: 'UPDATE_FIELD'; payload: { field: K; value: ContactFormFields[K]; error?: string } }
> | { type: 'START_FORM_SUBMISSION' }
> | { type: 'SET_FORM_SUBMISSION_SUCCESS' }
> | { type: 'SET_FORM_SUBMISSION_FAILURE'; payload: { error: string } }
> | { type: 'COMPLETE_FORM_SUBMISSION' }
> ```
> 
> This approach allows us to avoid dispatching unnecessary payloads when we don't need them.
> 
> For example, we only reset the form when it has been successfully submitted. To do that, we reset the fields to their initial values, set `isSubmitting` to `false`, and set `formSuccessfullySent` to `true` to indicate that the form was indeed sent. Since form validity is now computed on-the-fly, we don't have to manage it in the state 😉
> 
> Here's the reducer case:
> 
> ```tsx
> case 'SET_FORM_SUBMISSION_SUCCESS':
>   return {
>     ...state,
>     fields: initialContactFormFieldsState,
>     isSubmitting: false,
>     formSuccessfullySent: true,
>   };
> ```
> 
> I have also moved the timeout that resets the success button message into a `useEffect` hook. This ensures it is properly cleaned up if the component unmounts, preventing memory leaks and unintended behavior.
> 
> Let me know what you think about these changes! I'm happy to discuss any part of it 🙌 I'll be adding a few more comments soon with some additional improvements we can consider.

Yes, yours enhancements are pretty good! =)

Yura33-dev commented:
> @Yura33-dev Below are a few more suggestions to further enhance our form reducer and hook.
> 
> To keep the version history clean, could you please introduce these changes in three separate commits:
> 
> * Commit 1: Implement enums for action types.
> * Commit 2: Destructure state in useFormWithValidation.
> * Commit 3: Update fields validation (both email regex update and special characters restriction).
> 
> This way, each commit will have a clear purpose, making it easier to review and maintain 🤩
> 
> Definitely let me know if you have any questions or doubts!

No problem! I have pushed commits. But... with another commits in history 🙈

Yura33-dev committed:
refactor: implements enums for action types

Yura33-dev committed:
refactor: destructures state in useFormWithValidation

Yura33-dev committed:
refactor: updates fields validation

naumch1k commented:
@Yura33-dev, I’ve added the updated form error to [Figma](https://www.figma.com/design/YKZFEj97V6fe8LADQR5lcq/ONE-FALL?node-id=0-1&t=nmqX2YKfI9oxhQ0p-1). Could you implement it in our form? The text styles are consistent with `TextInput` and `TextArea`.
 
To avoid layout shifts when toggling the error text please ensure the error message maintains: one line height on desktop/tablets and two lines height on mobile.

Yura33-dev committed:
feature: implements form error

Yura33-dev commented:
@naumch1k Hey! I have added new commit with form error implementation. Could you please check it? Thanks!

Yura33-dev committed:
refactor: improves validation messages and regexp for validation

Yura33-dev committed:
refactor: changes validNameRegexp and validMessageRegexp

naumch1k committed:
feat: restyles form error to prevent layout shift

naumch1k committed:
feat: clears form error on resubmission

naumch1k commented:
@Yura33-dev I've added a few commits to prevent layout shifts on form error toggle and to clear error message on new submissions.

Previously, showing/hiding error messages caused layout shifts:

https://github.com/user-attachments/assets/81135115-e677-4879-b329-296d63f94cbc

https://github.com/user-attachments/assets/90efa52c-4ac7-4149-8fb9-1db6f43ced92

I removed conditional rendering for form error element and adjusted some styles to ensure it always occupies space in the footer layout to keep other elements stable when the error message appears or clears

These changes should enhance the user experience by providing a more consistent layout during form interactions. 

https://github.com/user-attachments/assets/fcfbf198-6ce2-4c70-8973-5a740ed5c46d

Please review the commits and let me know if you have any questions!

Yura33-dev commented:
> @Yura33-dev I've added a few commits to prevent layout shifts on form error toggle and to clear error message on new submissions.
> 
> Previously, showing/hiding error messages caused layout shifts:
> 
>  mobile.mov 
>  desktop.mov 
> I removed conditional rendering for form error element and adjusted some styles to ensure it always occupies space in the footer layout to keep other elements stable when the error message appears or clears
> 
> These changes should enhance the user experience by providing a more consistent layout during form interactions.
> 
>  desktop-after.mp4 
> Please review the commits and let me know if you have any questions!

I have reviewed your commits. Totally agree, they truly improves UX.  🙂

Yura33-dev committed:
refactor: edits validEmailRegexp according to RFC 5322 standard

Yura33-dev commented:
@naumch1k  Hi! I have added a new commit with edited `validEmailRegexp`. Now it allows to use ``!#$%&'*+-/=?^{|}~` in local parts of email field.

yuliia-harshyna reviewed (changes requested):
@naumch1k @Yura33-dev 
There are a few issues:

1. The symbols !#$&'*/=?^`{|}~ are not accepted by the email field.

2. The form isn`t cleared after the page reload (only in the Mozila Firefox browser, version 131.0).

3. The button 'SEND MESSAGE' is clickable after entering:

- '<>-+.@()' symbols and digits in the 'Name' field.

- Cyrillic in the subdomain part of the email (for Microsoft Edge browser, version 129.0.2792.79; Google Chrome browser, version 129.0.6668.90).

- first or last space in the 'Email' field (I'm not sure if this is a bug, because I don't know if it affects how the email is processed by the system).

- two dots consecutively in the domain name, first and last hyphens in the subdomain part of the 'Email' field.

- one Cyrilic letter as a first in the local part of the 'Email' field (two or more Cyrilic letters, Cyrilic letters in the middle or in the end of the local part are validated correctly).

- two dots consecutively in the local part of the 'Email' field;

- last dot in the domain part of the 'Email' field (so that email like '[test.@.email.com](mailto:test.@.email.com) is validated as a correct);

- the 'Message field' allows enter 500 characters maximum.

Also I suggest:

1. Mark all fields as required with '*' symbol. Because in case the field is empty, button is disabled and the user may not understand why he cannot submit the form.

2. Replace word 'letters' in the validation messages 'Message should have at least 2 letters' and 'Message cannot exceed 1000 letters' with the word 'characters'. Because this field allows enter not only letters.

naumch1k committed:
fix: repositions hyphen in name regex to prevent unintended ranges

naumch1k committed:
fix: changes form message limit to 1000 characters

naumch1k committed:
feat: adds '*' to form fields to indicate required inputs

naumch1k committed:
fix: improves validation messages for message form field

naumch1k committed:
fix: adjusts email regex

naumch1k commented:
@yuliia-harshyna 👋 I implemented some changes based on your feedback:
- Disallowed `!#$&'*/=?^`{|}~` for **name** field.
- Changed **message** field max characters limit to 1000 and replaced the word _letters_ with _characters_ in validation message.
- Added an asterisk (*) for form field placeholders to indicate required fields and prevent any user confusion.
- Did my best to incorporate all of your suggestions for email validation. This one was very tricky 🤯 It still allows Cyrillic letters in the domain part - and I have no idea how to fix it right now. Do you think we can leave it like that? 

Also, I was not able to reproduce the following:
> The form isn`t cleared after the page reload (only in the Mozila Firefox browser, version 131.0).

Would you be able to provide a more detailed description? Screen recording, maybe? (You can drop it right here in the comment).

ghost commented:
@naumch1k Hey!

- regarding the Cyrillic letters let`s leave it like that;

- here is the behavior of the site after reloading the page in the browser Mozila Firefox.
After reloading the page all fields in the form are displayed as filled. After typing value in any input field, the rest of the fields are cleared. But the submit button is still clickable.

https://github.com/user-attachments/assets/d2c02776-96a6-4072-bb98-40f955212525
